pkg/kube/client.go

```diff
@@ -377,18 +378,26 @@ func batchPerform(infos ResourceList, fn func(*resource.Info) error, errs chan<-
 }
 
 func createResource(info *resource.Info) error {
-	obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
-	if err != nil {
-		return err
-	}
-	return info.Refresh(obj, true)
+	return retry.RetryOnConflict(
```

Comment:
Hi,

Do you mean:
~~~
func createResource(info *resource.Info) error {
	return retry.RetryOnConflict(
		retry.DefaultRetry,
		func() error {
			return createResourceNoRetry(info)
		})
}

func createResourceNoRetry(info *resource.Info) error {
	obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
	if err != nil {
		fmt.Println("here")
		e, _ := json.Marshal(err)
		fmt.Println(string(e))
		return err
	}
	return info.Refresh(obj, true)
}
~~~

Or:
~~~
func batchPerform(infos ResourceList, fn func(*resource.Info) error, errs chan<- error) {
	var kind string
	var wg sync.WaitGroup
	for _, info := range infos {
		currentKind := info.Object.GetObjectKind().GroupVersionKind().Kind
		if kind != currentKind {
			wg.Wait()
			kind = currentKind
		}
		wg.Add(1)

		go func(info *resource.Info) {
			// Retry if this is a 409 Conflict
			errs <- retry.RetryOnConflict(
				retry.DefaultRetry,
				func() error {
					return fn(info)
				})
			wg.Done()
		}(info)
	}
}
~~~

If I add the change to batchPerform, there is a problem: `updateResource` calls `createResource` and `deleteResource` directly. It does not call `perform`. So, changing `createResource` and `deleteResource` would catch more issues than just chaning `batchPerform`.

Reply:
Personally, I'd prefer to update `createResource` and `deleteResource` directly. After some thought, `batchPerform` is not the correct place to address this. Not all methods (e.g. `Update`) call `batchPerform`. The only way to add this to `batchPerform` would be that all methods use `perform` / `batchPerform`, but that would require more changes than I intended. I hope that makes sense.

Reply:
Do you prefer this:
~~~
func createResource(info *resource.Info) error {
	return retry.RetryOnConflict(
		retry.DefaultRetry,
		func() error {
			return createResourceNoRetry(info)
		})
}

func createResourceNoRetry(info *resource.Info) error {
	obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
	if err != nil {
		fmt.Println("here")
		e, _ := json.Marshal(err)
		fmt.Println(string(e))
		return err
	}
	return info.Refresh(obj, true)
}
~~~

Over the current suggestion with the nested anonymous function? If so, I will amend the commit and revert to this.

Reply:
Hi,

Just to reiterate - the simpler implementation here [0] IMO is not the correct place to implement a retry. I think that a retry should be implemented in the individual createResource/deleteResource/updateResource methods because createResource/deleteResource are also individually called from `Update` (`Update` does not call `perform`, thus fixing this in `batchPerform` would not fix this for `Update`.

The current suggestion [1] addresses the issue at the source, in `createResource` and `deleteResource`. It also is IMO not complex. It's just a wrapper around the current logic. I could also imagine a scenario with a function `func retryResource(info *resource.Info, fn func (*resource.Info) error) which could be called everywhere like this:
`retryResource(info, createResource)`, `retryResource(updateResource)`. But that's actually more complex than [1] at this point in time.
In order to avoid further complexity, I also left out `updateResource`, as I dread that changing this one might introduce bugs or other complications and the issue with `resourcequota` shows up only when the actual quota changes.

The current unit tests show that [1] fixes what needed to be fixed. [0] Does not pass the new `TestUpdate` variation.
If you prefer [0], then I can recommit this (including a revert of `TestUpdate` and only using `TestCreate` to test this).

- Andreas 

[0]
~~~
func batchPerform(infos ResourceList, fn func(*resource.Info) error, errs chan<- error) {
	var kind string
	var wg sync.WaitGroup
	for _, info := range infos {
		currentKind := info.Object.GetObjectKind().GroupVersionKind().Kind
		if kind != currentKind {
			wg.Wait()
			kind = currentKind
		}
		wg.Add(1)

		go func(info *resource.Info) {
			// Retry if this is a 409 Conflict
			errs <- retry.RetryOnConflict(
				retry.DefaultRetry,
				func() error {
					return fn(info)
				})
			wg.Done()
		}(info)
	}
}
~~~

[1]
~~~
func createResource(info *resource.Info) error {
	return retry.RetryOnConflict(
		retry.DefaultRetry,
		func() error {
			obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
			if err != nil {
				return err
			}
			return info.Refresh(obj, true)
		})
}

func deleteResource(info *resource.Info) error {
	return retry.RetryOnConflict(
		retry.DefaultRetry,
		func() error {
			policy := metav1.DeletePropagationBackground
			opts := &metav1.DeleteOptions{PropagationPolicy: &policy}
			_, err := resource.NewHelper(info.Client, info.Mapping).DeleteWithOptions(info.Namespace, info.Name, opts)
			return err
		})
}
~~~

Reply:
@cndoit18 Any inputs?